### ISSUE TITLE
ComboBox.newItemProvider Issue when an new Item is set a second time

### ISSUE THREAD

**Issue by pinonpierre** [vaadin/framework#11317]
Hi,

comboBox.newItemProvider still not work correctly on Vaadin Framwork 8.6.0

The problem occurs when a new item is added a second time just after a dataProvider refresh that remove this newItem.

For instance, take this code:
```
        VerticalLayout mainLayout = new VerticalLayout();
        setCompositionRoot(mainLayout);

        List<String> items = new ArrayList<>(Arrays.asList("blue", "red", "green"));
        ListDataProvider<String> dataProvider = new ListDataProvider<>(items);

        ComboBox<String> comboBox = new ComboBox<>();
        comboBox.setDataProvider(dataProvider);
        comboBox.setNewItemProvider(value -> {
            items.add(value);
            dataProvider.refreshItem(value); //I don't know if it's required but it doesn't impact the bug
            return Optional.ofNullable(value);
        });
        mainLayout.addComponent(comboBox);

        Button reloadDataProviderButton = new Button("Reload Data Provider");
        reloadDataProviderButton.addClickListener(event1 -> {
            items.clear();
            items.addAll(Arrays.asList("blue", "red", "green"));
            dataProvider.refreshAll(); //I don't know if it's required but it doesn't impact the bug
        });
        mainLayout.addComponent(reloadDataProviderButton);

        Button getValueButton = new Button("Get Value");
        getValueButton.addClickListener(event1 -> event1.getButton().setCaption("Get Value: [" + comboBox.getValue() + "]"));
        mainLayout.addComponent(getValueButton);
```

And do these steps:
- Enter a new item
=> The field value equals this new item

- Clear the field
=> The field value is null

- Reload the data provider (The new value previously added is removed)

- Enter a second time the previous new item
=> The field value remains at null whereas should contains the new item

A quick gif to show you the issue with the above code:

![new_item_bug](https://user-images.githubusercontent.com/4507910/48482754-d7f44c80-e811-11e8-8ec0-7e735703b94f.gif)

Thanks,

**Comment by pinonpierre** [comment]
Hi,

Thanks for your answer but when you said that it's the best appoach, it's still a workaround and it's not the correct way to use it, right?

**Comment by evanzel** [comment]
Yes, I would prefer it if a reset() was not necessary, but currently it is.

**Comment by pinonpierre** [comment]
But I tested your fix and it doesn't work for me.

**Comment by pinonpierre** [comment]
This workaround doesn't seems to work @TatuLund if I replace:

`dataProvider.refreshItem(value) by comboBox.setSelectedItem(value)`

and

`dataProvider.refreshAll() by comboBox.getDataCommunicator().reset()`

**Comment by pinonpierre** [comment]
The problem is that the NewItemProvider listener is never call the second time

**Comment by evanzel** [comment]
Sorry, you are correct.
I was on 8.5.0 
Seems to be broken after 8.5.1.

**Comment by pinonpierre** [comment]
Thanks to remove the workaround tag @TatuLund

**Comment by ZheSun88** [comment]
related bug behavior:

after all the steps described, the ComboBox Popup cannot be opened.

**Comment by ZheSun88** [comment]
This issue is caused by the `lastnewItemString` in `VcomboBox`. 

This parameter is used to save the new added Items and it supposed to be replaced with the new item everytime. For our case, we are going to add the same item for the 2nd times, but `lastNewItemString` is not updated with the same `dataProvider.refresh()` call, so that the value doesn't trigger the client side update.

**Comment by ZheSun88** [comment]
This issue can be fixed by reverting the commit #11094, which will also fix issue #11320. 
The older issue #11059 needs to be verified.

**Comment by pinonpierre** [comment]
I don't understand because this bug is closed but I still have the issue in 8.6.2

**Comment by pinonpierre** [comment]
For Information: dataProvider.refreshAll() must be used instead of dataProvider.refreshItem(value) to see the fix work.
